### DISCUSSION TITLE
Assert when building

### DISCUSSION THREAD

**Discussion by jakob0** [carbon-language/carbon-lang#3492] [Getting started]
Hello.
I've been trying (and failing) to compile carbon for a while now.
I'm using OpenSUSE Tumbleweed and I've tried various versions of clang (14.0.0, 16.0.6, 17.0.4).
When I run:

`bazel run //explorer -- ./explorer/testdata/print/format_only.carbon`

bazel is always failing with an assertion:
```
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 8e4d337f-18d8-4579-9f00-77d45813ccb8
INFO: Analyzed target //explorer:explorer (70 packages loaded, 1750 targets configured).
INFO: Found 1 target...
ERROR: /home/jakob/Devel/test/carbon-lang/explorer/ast/BUILD:9:11: Compiling explorer/ast/clone_context.cpp failed: (Exit 1): clang++ failed: error executing command (from target //explorer/ast:ast) /home/jakob/.conan/data/llvm/16.0.6/dyn+conan/stable/package/c73a9a2d0216b447febbadf90db917614be4d518/bin/clang++ -no-canonical-prefixes -fcolor-diagnostics -Werror -Wall -Wextra -Wthread-safety ... (remaining 100 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
clang++: /builds/dyn/conan_llvm/.conan/data/llvm/16.0.6/dyn+conan/stable/build/c73a9a2d0216b447febbadf90db917614be4d518/llvm/clang/lib/Sema/SemaExprCXX.cpp:1154: clang::QualType adjustCVQualifiersForCXXThisWithinLambda(ArrayRef<clang::sema::FunctionScopeInfo *>, clang::QualType, clang::DeclContext *, clang::ASTContext &): Assertion `CurDC == getLambdaAwareParentOfDeclContext(CurLSI->CallOperator)' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/jakob/.conan/data/llvm/16.0.6/dyn+conan/stable/package/c73a9a2d0216b447febbadf90db917614be4d518/bin/clang++ -no-canonical-prefixes -fcolor-diagnostics -Werror -Wall -Wextra -Wthread-safety -Wself-assign -Wimplicit-fallthrough -Wctad-maybe-unsupported -Wdelete-non-virtual-dtor --system-header-prefix=external/ -c -MD -MF bazel-out/k8-fastbuild/bin/explorer/ast/_objs/ast/clone_context.pic.d -frandom-seed=bazel-out/k8-fastbuild/bin/explorer/ast/_objs/ast/clone_context.pic.o -std=c++17 -stdlib=libc++ -ffunction-sections -fdata-sections -fPIC -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" "-DCLANG_VERSION_FOR_CACHE=\"16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)\"" -DLLVM_ON_UNIX=1 -DHAVE_BACKTRACE=1 -DBACKTRACE_HEADER=<execinfo.h> -DLTDL_SHLIB_EXT=\".so\" -DLLVM_PLUGIN_EXT=\".so\" -DLLVM_ENABLE_THREADS=1 -DHAVE_DEREGISTER_FRAME=1 -DHAVE_LIBPTHREAD=1 -DHAVE_PTHREAD_GETNAME_NP=1 -DHAVE_PTHREAD_H=1 -DHAVE_PTHREAD_SETNAME_NP=1 -DHAVE_REGISTER_FRAME=1 -DHAVE_SETENV_R=1 -DHAVE_STRERROR_R=1 -DHAVE_SYSEXITS_H=1 -DHAVE_UNISTD_H=1 -D_GNU_SOURCE -DHAVE_LINK_H=1 -DHAVE_MALLINFO=1 -DHAVE_SBRK=1 -DHAVE_STRUCT_STAT_ST_MTIM_TV_NSEC=1 -DLLVM_NATIVE_ARCH=\"X86\" -DLLVM_NATIVE_ASMPARSER=LLVMInitializeX86AsmParser -DLLVM_NATIVE_ASMPRINTER=LLVMInitializeX86AsmPrinter -DLLVM_NATIVE_DISASSEMBLER=LLVMInitializeX86Disassembler -DLLVM_NATIVE_TARGET=LLVMInitializeX86Target -DLLVM_NATIVE_TARGETINFO=LLVMInitializeX86TargetInfo -DLLVM_NATIVE_TARGETMC=LLVMInitializeX86TargetMC -DLLVM_NATIVE_TARGETMCA=LLVMInitializeX86TargetMCA -DLLVM_HOST_TRIPLE=\"x86_64-unknown-linux-gnu\" -DLLVM_DEFAULT_TARGET_TRIPLE=\"x86_64-unknown-linux-gnu\" -DLLVM_VERSION_MAJOR=18 -DLLVM_VERSION_MINOR=0 -DLLVM_VERSION_PATCH=0 -DLLVM_VERSION_STRING=\"18.0.0git\" -D__STDC_LIMIT_MACROS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DLLVM_ENABLE_ZLIB=1 -DLLVM_ENABLE_ZSTD=1 -DZSTD_MULTITHREAD -DBLAKE3_USE_NEON=0 -DBAZEL_CURRENT_REPOSITORY=\"\" -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/llvm-project -iquote bazel-out/k8-fastbuild/bin/external/llvm-project -iquote external/llvm_zlib -iquote bazel-out/k8-fastbuild/bin/external/llvm_zlib -iquote external/llvm_zstd -iquote bazel-out/k8-fastbuild/bin/external/llvm_zstd -Ibazel-out/k8-fastbuild/bin/external/llvm_zlib/_virtual_includes/zlib -Ibazel-out/k8-fastbuild/bin/external/llvm_zstd/_virtual_includes/zstd -isystem external/llvm-project/llvm/include -isystem bazel-out/k8-fastbuild/bin/external/llvm-project/llvm/include -O1 -gmlt -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-optimize-sibling-calls -fsanitize=address,undefined,nullability -fsanitize-address-use-after-scope -fno-sanitize-recover=undefined -fsanitize-undefined-strip-path-components=-1 -fno-sanitize=vptr -D_LIBCPP_ENABLE_ASSERTIONS=1 -march=x86-64-v2 explorer/ast/clone_context.cpp -o bazel-out/k8-fastbuild/bin/explorer/ast/_objs/ast/clone_context.pic.o
```

Unfortunately, I have never used bazel before, so it's difficult for me to debug the problem on my own. Is my clang version missing some feature? Do you have any idea, where to look at? Any help would be appreciated.
Thanks
Jakob

**Comment by jonmeow** [comment]
This appears to be a clang crash, not bazel (last 4 lines of output).

On a quick search, I found https://github.com/llvm/llvm-project/issues/64359 -- apparently this has already been reported.  Unfortunately I don't reproduce the issue using 16.0.6 (Debian build), but that issue indicates that it may reproduce with LLVM head.

I see this is with //explorer, though. Most of the development focus at this point is on //toolchain. If you're just looking to run code you can still use the explorer at http://carbon.compiler-explorer.com. If you're looking to contribute, you might be able to just sidestep the build issue (clone_context.cpp is going to be doing some template code that's probably both the crash cause, and not used by toolchain code). If you have questions regarding the toolchain work, [#toolchain on Discord](https://discord.com/channels/655572317891461132/655578254970716160) is the best spot to ask right now.

**Comment by jakob0** [comment]
Thanks a lot for the answer.
Oh, I didn't find that issue. It looks like the same problem. Since I get the same assert error for different clang versions, i guess my build commands or my system are not really compatible with the carbon project.
You are right, I don't really care about the explorer. I just tried that, since it's the command suggested in the "getting started" section. When I try to run toolchain tests, I don't get that error (It still doesn't compile because of some include issues, but that's a different problem).

**Reply by jonmeow** [reply]
I'll go ahead and update those instructions to remove the reference.

My best guess is that there's some flag change due to the environment, but I'm not sure what it'd be. The crash is in clang's semantic analysis, so I'd expect it to be platform-independent. libc++ should be [fairly] platform-independent, but maybe there's some difference due to system headers? If you're able to narrow down a reduced repro (i.e., figure out some small-ish bit of code that reproduces) it might be easier to get traction on the LLVM issue.

Thinking about the libc++ bit, make sure you have a libc++-dev that's the same version as clang. I'd expect this to be true, but it's on my mind because recently there were some contributors hitting issues with version-locked libc++-dev installs as a consequence of updating clang. :)
